File: Refactoring/java/src/main/java/chapter10/Bird.java
package chapter10;

public class Bird {
    private String name;
    private String type;
    private Integer coconuts;
    private Integer voltage;
    private Boolean isNailed;

    public Bird(String name, String type, Integer coconuts, Integer voltage, Boolean isNailed) {
        this.name = name;
        this.type = type;
        this.coconuts = coconuts;
        this.voltage = voltage;
        this.isNailed = isNailed;
    }

    public String getName() {
        return name;
    }

    public String getType() {
        return type;
    }

    public Integer numberOfCoconuts() {
        return 5; // hard code
    }

    public Integer voltage() {
        return 120; // hard code
    }

    public Boolean isNailed() {
        return Boolean.TRUE; // hard code
    }
}

File: Refactoring/ts/bird.ts
interface IBird {
    name: string
    type: string
    numberOfCoconuts: number
    voltage: number
    isNailed: boolean
}

class Bird {
    bird: IBird

    constructor(bird: IBird) {
        this.bird = bird
    }

    get plumage(): string { // 깃털 상태
        return "알 수 없다";
    }
    
    get airSpeedVelocity(): number | null { // 비행속도
        return null;
    }
}
class EuropeanSwallow extends Bird {
    get plumage(): string { // 깃털 상태
         return '보통이다';
    }
    
    get airSpeedVelocity(): number | null { // 비행속도
        return 35;
    }
}
class AfricanSwallow extends Bird {
    get plumage(): string { // 깃털 상태
       return (this.bird.numberOfCoconuts > 2) ? "지쳤다" : "보통이다";
    }
    
    get airSpeedVelocity(): number | null { // 비행속도
       return 40 - 2 * this.bird.numberOfCoconuts;
    }
}
class NorwegianBlueParrot extends Bird {
    get plumage(): string { // 깃털 상태
       return (this.bird.voltage > 100) ? "그을렸다": "예쁘다";
    }
    
    get airSpeedVelocity(): number | null { // 비행속도
        return (this.bird.isNailed) ? 0: 10 + this.bird.voltage / 10;
    }
}